src/microbe_stage/BindingExtensions.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+public static class BindingExtensions
+{
+    private static Dictionary<Tuple<Microbe, string>, List<Action<object, object>>> actions =
+        new Dictionary<Tuple<Microbe, string>, List<Action<object, object>>>();
```

Comment:
Especially because there's state here.
If it is too much stuff to put in Microbe, I'd say you could at least make Microbe a partial class and put these in there.
But I think much of this should actually be in the colony class. It isn't really that much extra code needed in microbe to then do `Colony.Whatever` to perform operations.

Reply:
Putting this logic into the Colony is not possible, becuase this would not work if Colony is null.

Reply:
Then why does this need to exist when the colony doesn't exist?
I don't like this class in a design sense at all...

Reply:
Copying all of this code into the Microbe is possible and makes no difference logic wise, but this would make the (in my opinion already too long) Microbe.cs even longer.

Reply:
I really dislike having this global, static class holding a bunch of callbacks.
I'd really want to see a CellColony class created, that holds all of this state, and as much logic from here as possible.
I think it sohuld be then quite nice in the Microbe.cs to just have this kind of stuff dotted around:
```
if(Colony != null)
    Colony.Something();
```

Reply:
Just to make sure that we are both on the same level: do you know how extension methods work?
This allows me to call something like this: `myMicrobeInstance.GetAllColonyMembers()`.
Nobody will ever have to write the word `BindingExtensions`.

Reply:
I know how they work. I just don't think this class has any right to exist because it contains this very ugly global state.
I think that there is a reason why code design wise any Microbe needs to be able to perform colony operations directly like that. As I said in my previous comments, everything should go through the Colony property of a Microbe. So no need for extension methods or any proxies in Microbe.

Reply:
We're talking past each other.
How do you wish to get a list of all colony members, which returns one list entry if the cell is not in a colony?

Reply:
I want explicit code, the class name is Microbe. If some code needs special handling when a microbe is part of a colony it should be like
```
if(microbe.Colony != null){
then something like microbe.Colony.Members
} else {
microbe is just a single microbe
}
```

Reply:
I'm still confused.
Look at the base call in the `ColonyCompoundBag` class.
`public ColonyCompoundBag(Microbe microbe) : base(microbe.GetAllColonyMembers().Sum(p => p.Compounds.Capacity))`
how do you wan't this code to look like?

Reply:
I would like that code to be thrown out from Microbe. And only when a MicrobeColony object exists, then there is a corresponding ColonyCompoundBag, with code like:
```
public ColonyCompoundBag(MicrobeColony colony) : base(colony.Members.Sum(p => p.Compounds.Capacity))
```

Reply:
I don't think I've worded this quite good enough, so I'll try again now after some time.

Basically I want the design to be changed from:
Microbe, BindingExtensions, and ColonyCompoundBag being done and rewritten into the Microbe stuff, to
Microbe class has only minimal changes to call some methods in a Colony it is part of, Colony has the state of a colony as well as it will call methods on Microbe class to instruct it to do the right things, the colony will own the compound bag object.

So basically, I want the design to be based on a new MicrobeColony class that performs colony related operations. Microbe will be minimally changed with a new Colony property that is non-null when it is a part of a colony, and it will call some methods in the colony in places that need special handling when the cell is a part of a colony. Then all other code that needs to be colony aware will check microbe.Colony != null and call appropriate methods on the colony object.